### PR TITLE
IOCTL Interface for MsQuic Performance Counters

### PR DESCRIPTION
**Pull request by nibanks**
Adds an IOCTL interface that allows for user-mode clients to query the performance counters from msquic.sys. Currently untested.

### CODE REVIEW

File: src/bin/winkernel/ioctl.c [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    WDF_OBJECT_ATTRIBUTES Attribs;
+    WDFDEVICE Device;
+    QUIC_DEVICE_EXTENSION* DeviceContext;
+    WDF_IO_QUEUE_CONFIG QueueConfig;
+    WDFQUEUE Queue;
+
+    DeviceInit =
+        WdfControlDeviceInitAllocate(
+            Driver,
+            &SDDL_DEVOBJ_SYS_ALL_ADM_ALL);
```
Comment (nibanks) [inline src/bin/winkernel/ioctl.c:73]: Who should have access to query these performance counters?

File: src/manifest/clog.sidecar [hunk 2]
```diff
@@ -24637,6 +24637,34 @@
       {
         "UniquenessHash": "977ca100-03bb-010a-09ab-0a04d33a7dfd",
         "TraceID": "DatapathDropAllocRecvBufferFailure"
+      },
+      {
+        "UniquenessHash": "fd2e7ffd-0428-e1d7-902e-d4ff393b53bb",
```
Comment (nibanks) [inline src/manifest/clog.sidecar:24642]: Clog is still only adding the hashes for kernel mode.

File: src/bin/winkernel/ioctl.c (hunk 1 above)
Comment (ThadHouse) [reply src/bin/winkernel/ioctl.c:73]: Perfmon doesn't force admin, so I think it needs to be all users?

**Review by ThadHouse — APPROVED** [review]

**Review by ThadHouse — APPROVED** [review]

File: src/bin/winkernel/ioctl.c [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        //
+        // Copy as many counters will fit completely in the buffer.
+        //
+        OutputBufferLength = (OutputBufferLength / sizeof(int64_t)) * sizeof(int64_t);
+        if (OutputBufferLength == 0) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Error;
+        }
+    }
+    else {
```
Comment (nibanks) [inline src/bin/winkernel/ioctl.c:356]: formatting

File: src/inc/msquic_ioctl.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            0,
+            (uint8_t*)Counters,
+            (*NumberOfCounters) * 8,
+            &ReadBytes,
+            NULL)) {
+        Status = HRESULT_FROM_WIN32(GetLastError());
+        goto Exit;
+    }
+
+    *NumberOfCounters = (uint32_t)(ReadBytes / 8);
```
Comment (nibanks) [inline src/inc/msquic_ioctl.h:86]:
```suggestion
    *NumberOfCounters = (uint32_t)(ReadBytes / sizeof(int64_t));
```

File: src/inc/msquic_ioctl.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        CreateFileW(
+            L"\\\\.\\\\" MSQUIC_DEVICE_NAME,
+            GENERIC_READ,
+            0,
+            NULL,
+            OPEN_EXISTING,
+            FILE_ATTRIBUTE_NORMAL,
+            NULL);
+    if (DeviceHandle == INVALID_HANDLE_VALUE) {
+        Status = HRESULT_FROM_WIN32(GetLastError());
```
Comment (nibanks) [inline src/inc/msquic_ioctl.h:69]: Don't call `HRESULT_FROM_WIN32`.

File: src/inc/msquic_ioctl.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    This file contains the definitions for the msquic ioctl interface
```
Comment (nibanks) [inline src/inc/msquic_ioctl.h:8]:
```suggestion
    This file contains the definitions for the MsQuic IOCTL interface
```

File: src/inc/msquic_ioctl.h [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+#include "winioctl.h"
+#endif
+
+#ifdef QUIC_PRIVATE_INTERFACE
+#define MSQUIC_DEVICE_NAME L"msquicpriv"
+#else
+#define MSQUIC_DEVICE_NAME L"msquic"
+#endif
+
+
```
Comment (nibanks) [inline src/inc/msquic_ioctl.h:34]: nit: remove extra newline

File: scripts/test.ps1 [hunk 8]
```diff
@@ -193,11 +193,13 @@ $RunTest = Join-Path $RootDir "scripts/run-gtest.ps1"
 $MsQuicTest = $null
 $MsQuicCoreTest = $null
 $MsQuicPlatTest = $null
+$MsQuicIoctlTest = $null
 $KernelPath = $null;
 if ($IsWindows) {
     $MsQuicTest = Join-Path $RootDir "\artifacts\bin\windows\$($Arch)_$($Config)_$($Tls)\msquictest.exe"
     $MsQuicCoreTest = Join-Path $RootDir "\artifacts\bin\windows\$($Arch)_$($Config)_$($Tls)\msquiccoretest.exe"
     $MsQuicPlatTest = Join-Path $RootDir "\artifacts\bin\windows\$($Arch)_$($Config)_$($Tls)\msquicplatformtest.exe"
+    $MsQuicPlatTest = Join-Path $RootDir "\artifacts\bin\windows\$($Arch)_$($Config)_$($Tls)\msquicioctltest.exe"
```
Comment (nibanks) [inline scripts/test.ps1:202]:
```suggestion
    $MsQuicIoctlTest = Join-Path $RootDir "\artifacts\bin\windows\$($Arch)_$($Config)_$($Tls)\msquicioctltest.exe"
```

File: src/inc/quic_driver_helpers.h [hunk 9]
```diff
@@ -139,6 +139,213 @@ class QuicDriverService {
         }
         return true;
     }
+
+    VOID DoStopSvc()
```
Comment (nibanks) [inline src/inc/quic_driver_helpers.h:143]: Style in the function isn't consistent with our other code.

Comment (ThadHouse) [reply src/inc/quic_driver_helpers.h:143]: Yeah I know, I plan on digging through that. Its just a straight copy right now of the sample code.

File: src/test/ioctltestbin/main.cpp [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+#include "gtest/gtest.h"
+
+#ifdef QUIC_CLOG
+#include "main.cpp.clog.h"
+#endif
+
+extern "C" _IRQL_requires_max_(PASSIVE_LEVEL) void QuicTraceRundown(void) { }
+
+
+
```
Comment (nibanks) [inline src/test/ioctltestbin/main.cpp:28]: nit: too many newlines

**Pull request merged**